client.go

```diff
@@ -304,9 +316,60 @@ func (client *client) GetOffset(topic string, partitionID int32, time int64) (in
 	return offset, err
 }
 
+func (client *client) Coordinator(consumerGroup string) (*Broker, error) {
+	client.lock.RLock()
+	coordinator := client.coordinators[consumerGroup]
+	client.lock.RUnlock()
+
+	if coordinator == nil {
+		var err error
+		err = client.RefreshCoordinator(consumerGroup)
+		if err != nil {
+			return nil, err
+		}
+
+		client.lock.RLock()
+		coordinator = client.coordinators[consumerGroup]
+		client.lock.RUnlock()
+	}
+
+	_ = coordinator.Open(client.conf)
+	return coordinator, nil
+}
+
+func (client *client) RefreshCoordinator(consumerGroup string) error {
+	coordinator, err := client.getCoordinator(consumerGroup, client.conf.Metadata.Retry.Max)
+	if err != nil {
+		return err
+	}
+
+	client.lock.Lock()
+	client.coordinators[consumerGroup] = client.registerBroker(coordinator)
```

Comment:
OK, I feel like it's just over-complicated to return the result, and then write it to the cache in a separate method, when you could simply write it to the cache in the first place. getCoordinator's result is never going to be used for anything but updating the cache, so splitting that apart seems over-engineered to me.

We agree on Coordinator/Leader/etc should just read from the cache. I'm just not sure why we need two separate methods (fetch from remote, update cache) when that feels like a single operation to me.

More specifically: we don't want anybody to fetch from the cluster without updating the cache. What good would that do? Splitting the two methods gives us that flexibility, yes, but I think that's flexibility we should be avoiding. The cache is the source of truth. There is a single black box method that updates the cache.